test/quantization/pt2e/test_quantize_pt2e.py

```diff
@@ -1712,3 +1712,106 @@ def test_reentrant(self):
         self.checkGraphModuleNodes(
             m, expected_node_occurrence=node_occurrence, expected_node_list=node_list
         )
+
+    def test_observer_callback(self):
+        from torch.library import Library, impl
+        test_lib = Library("test_int4", "DEF")
+        test_lib.define("quantize_per_tensor_int4(Tensor input, float scale, int zero_point) -> Tensor")
+
+        @impl(test_lib, "quantize_per_tensor_int4", "CompositeExplicitAutograd")
+        def quantize_per_tensor_int4(
+            input: torch.Tensor,
+            scale: float,
+            zero_point: int,
+        ) -> torch.Tensor:
+            inv_scale = 1.0 / scale
+            return torch.clamp(torch.round(input * inv_scale) + zero_point, 0, 15).to(torch.uint8).view(torch.bits8)
+
+        test_lib.define("dequantize_per_tensor_int4(Tensor input, float scale, int zero_point) -> Tensor")
+
+        @impl(test_lib, "dequantize_per_tensor_int4", "CompositeExplicitAutograd")
+        def dequantize_per_tensor_int4(
+            input: torch.Tensor,
+            scale: float,
+            zero_point: int,
+        ) -> torch.Tensor:
+            return (input.view(torch.uint8).to(torch.float32) - zero_point) * scale
+
+        from torch.ao.quantization.observer import ObserverBase
+
+        class Int4Observer(ObserverBase):
+            def __init__(self, *args, **kwargs):
+                # just faking a dtype here
+                super().__init__(dtype=torch.int8)
+
+            def forward(self, x):
+                return x
+
+            def calculate_qparams(self, **kwargs):
+                pass
+
+            def convert(self, model: torch.fx.GraphModule, observer_node: Node):
```

Comment:
I would prefer us to have `convert` do only specific things related to observer.

We need to figure out what this thing is?

My proposal is this:
Observer's forward method takes in a tensor and returns another tensor. Convert should just return equivalent of that but with whatever transforms they want to apply but not really doing any graph mutation.

e.g.
```
def convert(self, x):
   x_q = torch.ops.test_int4.quantiize_per_tensor_int4(...)
   return torch.ops.test_int4.dequantize_per_tensor_int4(x_q)
```

It is quant workflow's responsibility to replace original instance of observer with equivalent graph of convert. 

I would like to avoid 1) having observer to have think about both eager mode behavior (forward method) but also graph mode behavior for convert and 2) having to pass the entire GraphModule to convert which is error prone and allows for arbitrary mutation on the graph.

Reply:
yeah I thought about this, this is nice if we can achieve it, but I think it's hard to do because we also need to add new attributes to the graph module as well, what I wrote in the test was a simplest version that skipped a lot of details.

Reply:
I think we should discuss in detail what we want to achieve here first though and then list constraints

Reply:
I think we should do that first before landing. WDYT?

Reply:
yes sure, will follow up with you offline